go/decompress/decompress_test.go

```diff
@@ -0,0 +1,349 @@
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		testReadLinesWithBuffer(100_000, 400)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		testReadLinesWithBuffer(100_000, 1000)
+	}()
+
+	wg.Wait()
+}
+
+func testReadLinesWithBuffer(numLines int, bufferSize int) {
+	// Step 1: Generate Test Data
+	testDataBytes := generateTestData(numLines)
+
+	// Step 2: Compress the Data
+	random_name := randStringBytes(10)
+	compressedFileName := fmt.Sprintf("test_data_%s.bgz", random_name)
+	compressData(testDataBytes, compressedFileName)
+
+	// Step 3: Read the Compressed Data
+	expectedLines := strings.Split(string(testDataBytes), "\n")
+	testDataBytes = nil
+
+	err := readCompressedData(compressedFileName, expectedLines, bufferSize)
+
+	// delete the file
+	err2 := os.Remove(compressedFileName)
+	if err != nil {
+		log.Fatalf("Error reading compressed data: %v", err)
+	}
+
+	if err2 != nil {
+		log.Fatalf("Error deleting compressed file: %v", err2)
+	}
+}
+
+func generateTestData(numLines int) []byte {
+	min := 1
+	max := 3000
+
+	var buffer bytes.Buffer
+	for i := 0; i < numLines; i++ {
+		// generate a random line
+
+		buffer.WriteString(fmt.Sprintf("Line %d %s\n", i, randStringBytes(rand.Intn(max-min)+min)))
+	}
+	return buffer.Bytes()
+}
+
+func compressData(data []byte, fileName string) {
+	f, _ := os.Create(fileName)
+	defer f.Close()
+
+	w := bgzf.NewWriter(f, 1)
+	defer w.Close()
+
+	w.Write(data)
+}
+
+func readCompressedData(fileName string, expectedLines []string, bufferSize int) error {
+	f, _ := os.Open(fileName)
+	defer f.Close()
+
+	r, _ := bgzf.NewReader(f, 0)
+	defer r.Close()
+
+	i := 0
+	for {
+		lines, err := readLinesBgzipWithBuffer(r, bufferSize)
+
+		if len(lines) > 0 {
+			strLines := strings.Split(string(lines), "\n")
+
+			for _, line := range strLines {
+				if line != expectedLines[i] {
+					return fmt.Errorf("Line %d does not match. Got: %s, expected: %s", i, line, expectedLines[i])
+				}
+
+				i += 1
+			}
+		}
+
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+	}
+}
+
+// NOTE: test data comes from `../test/opensearch/testdata/input.txt`
+func Test_readLinesWithBuffer(t *testing.T) {
+	type args struct {
+		bufferSize int
+		input      []byte
+	}
+	type want struct {
+		file   string
+		golden []byte
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    []byte
+		wantErr error
+	}{
+		{
+			name: "empty input",
+			args: args{
+				bufferSize: 100,
+				input:      []byte(""),
+			},
+			want:    []byte(""),
+			wantErr: io.EOF,
+		},
+		{
+			name: "negative buffer",
+			args: args{
+				bufferSize: -1,
+				input:      []byte(""),
+			},
+			want:    []byte(""),
+			wantErr: io.EOF,
+		},
+		{
+			name: "ends with \\n",
+			args: args{
+				bufferSize: 100,
+				input:      []byte("a\tb.c\td.e.f\n1\tA\t1;2|3;4/5\n1|2\tA\t1;2|3;4/5\na/2\tA;B\t1/2|3;4/5\n"),
+			},
+			want:    []byte("a\tb.c\td.e.f\n1\tA\t1;2|3;4/5\n1|2\tA\t1;2|3;4/5\na/2\tA;B\t1/2|3;4/5"),
```

Comment:
I forgot to mention that the input and expected output don't match in this case (because I was playing around with it and I wasn't sure if this was your goal).  I would have expected the output has a final \n but it doesn't. Can you check on this?

Reply:
The read functions always trim the trailing new line. I’ll add a comment above the interface explaining that.